### DISCUSSION TITLE
Is my understanding of up-keep correct?

### DISCUSSION THREAD

**Discussion by thetutlage** [unpoly/unpoly#293] [Q&A]
I was following the `up-keep` guide that shows an audio player example (exactly what I am trying to do). However, when using the browser back button, the player resets even though the parent div has `up-keep` attribute on it.

Here's a glitch example of the same 

Preview - https://noiseless-upbeat-gray.glitch.me/index.html
Code: https://glitch.com/edit/#!/noiseless-upbeat-gray

## Steps to reproduct

- Play the audio on index.html page
- Press Page B link ( audio continue to play. Which is great )
- Press browser back link ( audio stops )

**Comment by adam12** [comment]
Your understanding is correct. This looks like a regression.

See my remix with an older version of 2.x which works correctly. https://glitch.com/edit/#!/daily-dear-zebu?path=index.html%3A7%3A61

**Comment by adam12** [comment]
Actually, my mistake, it's not necessarily a regression, but maybe an undocumented consequence of #237.

You can probably fix this by adding `up-main` back as the default restore target:

```js
up.history.config.restoreTargets.unshift('[up-main]')
```

See this remix: https://glitch.com/edit/#!/quilt-olivine-whip

**Reply by thetutlage** [reply]
Ahhh. That seems to work. I looked up the docs for `restoreTargets`, but there is very little information available on it. Is it possible for you to explain how exactly it works under the hood? To be specific.

- Does the array stops after the first match?
- And how does restoring works in general? Does unpoly claim the markup within the matched `restoreTarget` and then performs the same update logic that happens during normal click? Maybe, this is a bigger topic to explain and I would love to see some docs around it
- If there are multiple `up-main` in the document, which one will it match?

**Reply by thetutlage** [reply]
Okay, I have to say. I am still a little confused with `up-keep`. In the same glitch example, I have removed `up-main` from the section element and now `up-keep` resets even during standard navigation (not the back button).

I expect it to work similar to `data-turbo-permanent` from hotwire turbo. Here's a glitch example of hotwire turbo

- Demo: https://turbo-app.glitch.me/index.html
- Code: https://glitch.com/edit/#!/turbo-app

**Reply by adam12** [reply]
It used to work as you expected, but it looks like it changed/broke at version 0.35.0.

I think there's a few things going on:

1. When `body` is the target, it seems to not work correctly (since 0.35.0).
2. When `body` is the default restore target, it doesn't appear to restore correctly (maybe related).

I'd have to bisect that release to see if we can find the breakage, but maybe Henning would recognize it by the release. We'll have to see.

**Comment by triskweline** [comment]
> However, when using the browser back button, the player resets even though the parent div has up-keep attribute on it.

Preservation of `[up-keep]` elements was explicitly disabled when going back in history.

I don't really remember the reason for this, except maybe "reset more".

I just changed this in b67983afcb4a90225db55e8e3855c93c2a14d933 (not yet released).

> I have removed up-main from the section element and now up-keep resets even during standard navigation (not the back button).

When you use devtools to add a style or property to the `<audio>` element, you see that the element is preserved. However, something causes the player to stop playback.

I'm not sure why this happens. The code used to swap the `<body>` is the same code used to swap other kinds of fragments:

- Parse the server's HTML into a new document
- Transfer [up-keep] elements to the new document
- Swap old and new fragment. In this case `document.documentElement.replaceChild(newBody, oldBody)`.

I saw in your Glitch that Turbo can swap the body while keeping the audio element playing. Maybe it uses a different way to swap elements that we can adapt?

Regarding your questions about `up.history.restoreTargets`:

> Does the array stops after the first match?

Yes. If you need to update multiple elements use a comma-separated selector within a single array element, or use `[up-hungry]`.

> And how does restoring works in general? Does unpoly claim the markup within the matched restoreTarget and then performs the same update logic that happens during normal click?

It requests the URL from the previous history state and then replaces the first matching selector from `up.history.restoreTargets`.

> If there are multiple up-main in the document, which one will it match?

The first element will match. But [like with a HTML `<main>` element](https://html.spec.whatwg.org/multipage/grouping-content.html#the-main-element) you should only have one. If you need to replace a periphal element outside of the main element, like a page-specific sidebar, use `[up-hungry]`.

**Reply by adam12** [reply]
> I saw in your Glitch that Turbo can swap the body while keeping the audio element playing. Maybe it uses a different way to swap elements that we can adapt?

This did work correctly in 0.34.2. I noticed there's a new mechanism that uses `DOMParser` in that release and I wonder if it's related? I didn't get a chance yet to bisect 0.35 to see what broke.

**Reply by triskweline** [reply]
> This did work correctly in 0.34.2. I noticed there's a new mechanism that uses DOMParser in that release and I wonder if it's related?

DOMParser produces elements in a second `Document`, which can have an effect, but it's super rare. E.g. when you parse a `<script>` tag with DOMParser and insert it into your main document, the script tag is not executed because the spec says so.

What this doesn't explain is why the issue only appears when the `<body>` is swapped, but not for any other container element.

[Turbo also uses `DOMParser`](https://github.com/hotwired/turbo/blob/3c726f7087365dd77221ea51b5e6aa523004507f/src/util.ts#L28).

**Comment by triskweline** [comment]
I found this Turbo issue / PR regarding preservation of playback state: https://github.com/hotwired/turbo/pull/226

It seems they've switched from using `Element#replaceWith()` to `Element#replaceChild()`. I don't know if the rest of the diff is relevant to this issue. We would need to test this in different browsers to see.

**Comment by triskweline** [comment]
I created an issue for audio elements losing playback state when rendering the body:
https://github.com/unpoly/unpoly/issues/294

**Reply by hpvd** [reply]
audio elements losing playback state when rendering the body seems to be solved https://github.com/unpoly/unpoly/issues/294
